https://issues.redhat.com/browse/ACM-22475 Backup and restore Observability (#8222)
open-cluster-management/rhacm-docs: business_continuity/backup_restore/backup_restore_obs.adoc

[#backup-restore-obs]
= Backing up and restoring Observability service

> Backup and restore the Observability service to keep data safe and to maintain its continuity during the hub cluster migration or backup. To help with disruption in metric data collection, use the same S3-compatible object store for both the primary and backup hub clusters.

Reviewer: ```suggestion
Backup and restore the Observability service to keep data safe and to maintain continuity during the hub cluster migration or backup. To help with disruption in metric data collection, use the same S3-compatible object store for both the primary and backup hub clusters.
```

Reviewer: removing possessive

> Backup and restore the Observability service to maintain data continuity and integrity during hub cluster migration or backup. It is assumed that the same S3-compatible object store is used across both the primary and backup hub clusters to minimize disruption in metric data collection.

Reviewer 2: If you liked my earlier suggestions, you can apply the same suggestions here. 

Also, are we wanting the user to use the same S3-compatible object store? Or does Observability assume the user uses the same S3 compatible object store? 

If so, consider something like the following: 

```suggestion
Backup and restore the Observability service to keep data safe and to maintain its continuity during the hub cluster migration or backup. To help with disruption in metric data collection, use the same S3-compatible object store for both the primary and backup hub clusters.
```

Author: That is more clear, made a couple adjustments to your comment.

.Prerequsites

> - Ensure that you can run a restore operation for backup types by completing xref:../business_continuity/backup_restore/backup_restore.adoc#restoring-backup-restore-operation[Using the restore operation for backup types].

Reviewer: ```suggestion
- Ensure that you can run a restore operation for backup types by completing the xref:../business_continuity/backup_restore/backup_restore.adoc#restoring-backup-restore-operation[Using the restore operation for backup types] process.
```

[#backup-restore-obs-procedure]
== Backing up and restoring Observability procedure

> . To ensure the Observability service recognizes the hub cluster as the `local-cluster`, change the `spec.disableHubSelfManagement` parameter in the `MultiClusterHub` custom resource to `false`. 

Reviewer 2: I know Brandi wants us to use "managed hub cluster" instead of "local hub cluster" However, "local-cluster" might be different, and there are times when I've had "local-cluster" be a value in a yaml file, so I couldn't change the name. 

Nonetheless, might be worthwhile to confirm that it is best to use "local-cluster" here instead of "a managed cluster"

Author: @subbarao-meduri can you help confirm if my rewrite here makes sense and is accurate.

Reviewer 3: Hub cluster metrics are always collected and pushed into Thanos regardless of whether hub self management is enabled or not. See: https://docs.redhat.com/en/documentation/red_hat_advanced_cluster_management_for_kubernetes/2.14/html-single/observability/index#obs-config

These metrics show up under the name `local-cluster` in Grafana by default. However, if the `local-cluster` is [renamed](https://docs.redhat.com/en/documentation/red_hat_advanced_cluster_management_for_kubernetes/2.14/html-single/install/index#disable-hub-self-management) to something else, metrics would show up under that name. 

@dockerymick I don't believe there is an explicit requirement that hub self-management must be enabled for backup/restore to work. @birsanv can confirm this. My view is that line 13 should be removed.

cc: @coleenquadros to recheck my assertion here.

Author: @subbarao-meduri Thanks Subbarao for your feedback. This line was an attempt to rewrite what you mentioned in the issue description: 

"For ACM 2.14 and later: Use the local-cluster renaming feature to assign unique names to each hub. This helps disambiguate metrics for each hub cluster in Grafana." 

This info is not listed in the Prerequisistes portion. It is a part of the "Backing up and restoring Observability procedure " section

Reviewer: I think there is some confusion (maybe), I never said _not_ to use `local cluster`; it is better to use that if you are referring to a _managed hub_. I don't know what we mean by _local hub cluster_. I may be missing something--I am just not aware of that term being part of the product lexicon, such as hub cluster, managed cluster, local cluster.

This discussion originally stemmed from topics incorrectly or vaguely referring to the _local cluster_ because I think the team didn't have the same understanding of what local cluster means or how to disable and enable.

When we speak of _local cluster,_ or enabling and disabling it, we just need to make sure we understand ourselves how to do it, then present it the same as in the _adv. config,_ which is what this is line is doing from what I can tell.

If you talk about a _managed hub cluster,_ you should ask if they mean `local-cluster` and present it as such. That was my request for the team. I think this is fine, I would just have said:

_To ensure the Observability service recognizes the hub cluster as a `local-cluster`, a managed hub cluster, set the `spec.disableHubSelfManagement` parameter in the `MultiClusterHub` custom resource to `false`._ 

@jc-berger @dockerymick @subbarao-meduri FYI. That is if we don't remove this line.

Reviewer: We do also now talk about the ability to rename the local cluster, so if you keep that in there, you may want to say:

If you change the default name of `local-cluster` to another value, the results will appear within the changed local cluster name.

Reviewer 4: @dockerymick Subbarao's assertion on the metrics being collected under the new name for the managed hub or local cluster is correct. Also yes, the hub cluster will continue to send metrics regardless of HubSelfManagement being enabled or not

> . Manually back up and restore the `observatium` resource and `observability` deployment to ensure continuity across hub clusters.

Reviewer: Is this a step and below is sub steps for that step? Or can we merge this? Technically the first step here doesn't have an action. Can the step under this stand alone and this just be a paragraph?

Author: Initially I went with the substep because the step was becoming lengthy. However, I did make an update to remove the substep (i am just going through the comments/suggestions and updating the branch locally). I changed it to read as the following statement: 

"To preserve the tenant ID of the `observatorium` resource as you manually back up and restore the `observatorium` resource, run the following command:"

.. To preserve the tenant ID of the `observatorium` resource during the restore, run the following command:

+
[source,bash]
----
oc get observatorium -n open-cluster-management-observability -o yaml > observatorium-backup.yaml
----

. To backup the `observability` deployment, run the following command:

+
[source,bash]
----
oc get mco observability -o yaml > mco-cr-backup.yaml
----

> . Shut down the Thanos compactor on your primary hub cluster.

Reviewer: Also not a step bc there is not action. Can this be just a line to introduce the process?

Reviewer: Or add ...complete the following steps and you can keep it as a step because there is an action.

.. To prevent write conflicts and deduplication issues while working on the same object storage, stop the Thanos compactor before starting the restore operation on the backup hub cluster. Run the following command:

+
[source,bash]
----
oc scale statefulset observability-thanos-compact -n open-cluster-management-observability --replicas=0
----

.. Verify the compactor is stopped by running the following command:

+
[source,bash]
----
oc get pods observability-thanos-compact-0 -n open-cluster-management-observability
----

> . To restore the `backup` resources, see xref:../business_continuity/backup_restore/backup_restore.adoc#restoring-backup-restore-operation[Using the restore operation for backup types]. You can restore the automatically backed-up ConfigMaps and Secrets listed in the backup and restore configuration for Observability.

Reviewer 2: I'm not used to seeing links to other procedures/docs in the middle of a procedure/task. Could this link be a prereq? I know we might only need to restore the backup resoruces here in step 5, but maybe we can have a Prereq like this: 

Prerequisites 

- Ensure that you can run a restore operation for backup types by completing xref:../business_continuity/backup_restore/backup_restore.adoc#restoring-backup-restore-operation[Using the restore operation for backup types]. 

Then the step here can be simple like, 

5. Restore the `backup` resources, including: the automatically backed-up ConfigMaps and Secrets listed in the backup and restore configuration for Observability.

Author: Great point and callout for sure

. To preserve the tenant ID for maintaing continuity in the metrics ingestion and querying, restore the `observatorium` resource to the backup hub cluster. Run the following command:

+
[source,bash]
----
oc apply -f observatorium-backup.yaml
----

. Apply the backed up `MultiClusterObservability` custom resource to start the Observability service on the new restored hub cluster. Run the following command:

+
[source,bash]
----
oc apply -f mco-cr-backup.yaml
----
+
The operator starts the Observability service and detects the existing `observatorium` resource, reusing the preserved tenant ID instead of creating a new one.

> . Migrate managed clusters to the new hub cluster.

Reviewer: ```suggestion
. Migrate managed clusters to the new hub cluster. Complete the following steps:
```

Reviewer: Or let it stand alone without a step:

Migrate managed clusters to the new hub cluster. Complete the following steps:

.. Detach managed clusters from the primary hub cluster and reattach the managed clusters to the new restored hub cluster. After attaching the managed clusters, the clusters resume sending metrics to the Observability service.

> . Shut down Observability on your primary hub cluster and clear in-memory metrics to S3 object after migrating all managed clusters. Run the following command:

Reviewer: ```suggestion
. Shut down the Observability service on your primary hub cluster and clear in-memory metrics to S3 object after migrating all managed clusters. Run the following command:
```

[#backup-restore-obs]
= Backing up and restoring Observability service

Backup and restore the Observability service to keep data safe and to support continuity during the hub cluster migration or backup. To help with disruption in metric data collection, use the same S3-compatible object store for both the primary and backup hub clusters.

.Prerequsites

- Ensure that you can run a restore operation for backup types by completing the xref:../business_continuity/backup_restore/backup_restore.adoc#restoring-backup-restore-operation[Using the restore operation for backup types] process.

[#backup-restore-obs-procedure]
== Backing up and restoring Observability procedure

. To ensure the Observability service recognizes the hub cluster as the `local-cluster`, a managed hub cluster, change the `spec.disableHubSelfManagement` parameter in the `MultiClusterHub` custom resource to `false`.

+

> *Note:* If you change the default name of your `local-cluster` to another value, the results ppear within the changed local cluster name.

Reviewer 4: typo for appear

> *Note:* If you change the default name of your `local-cluster` to another value, the results ppear within the changed local cluster name.

Author: ```suggestion
*Note:* If you change the default name of your `local-cluster` to another value, the results appear within the changed local cluster name.
```

----

. To backup the `observability` deployment, run the following command:

+
[source,bash]
----
oc get mco observability -o yaml > mco-cr-backup.yaml
----

. Shut down the Thanos compactor on your primary hub cluster by running the following command:

+
[source,bash]
----
oc scale statefulset observability-thanos-compact -n open-cluster-management-observability --replicas=0
----

.. Verify the compactor is not active by running the following command:

+
[source,bash]
----
oc get pods observability-thanos-compact-0 -n open-cluster-management-observability
----

. Restore the `backup` resources such as the automatically backed-up ConfigMaps and Secrets listed in the backup and restore configuration for Observability.

. To preserve the tenant ID for maintaing continuity in the metrics ingestion and querying, restore the `observatorium` resource to the backup hub cluster. Run the following command:

+
[source,bash]
----
oc apply -f observatorium-backup.yaml
----

. Apply the backed up `MultiClusterObservability` custom resource to start the Observability service on the new restored hub cluster. Run the following command:

> 

Reviewer 3: @dockerymick add a new step:

Verify the ManagedClusters are Connected to the new ACM Hub after the restore is complete.  Verify that all ManagedClusters showing as "Available".

+
[source,bash]
----
oc apply -f mco-cr-backup.yaml
----
+
The operator starts the Observability service and detects the existing `observatorium` resource, reusing the preserved tenant ID instead of creating a new one.

> . Migrate managed clusters to the new hub cluster by detaching managed clusters from the primary hub cluster,  and reattach the managed clusters to the new restored hub cluster. After attaching the managed clusters, the clusters resume sending metrics to the Observability service.

Reviewer 3: @dockerymick A customer may not need to manually detach and re-attach their clusters but the backup restore process could do this for them. Please remove "Migrate managed clusters..." statement.

> 

Reviewer 3: @dockerymick add step

Verify observability stack is running on the new hub
```
"oc get pods -n open-cluster-management-observability"
```

Verify that `Observability-controller` managedclusteraddon on managed clusters is not in `degraded` state or Progressing state is not set to `false`
```
"oc get managedclusteraddons -A | awk 'NR==1 || /observability-controller/'"
```

Author: What does "Progressing state" look like? Is that a parameter key?

Reviewer 3: The DEGRADED column is empty, PROGRESSING column should say `False`. Example below.

```
# oc get managedclusteraddons -A  | awk 'NR==1 || /observability-controller/'
NAMESPACE                               NAME                          AVAILABLE   DEGRADED   PROGRESSING
...
smeduri-spoke                           observability-controller      True                   False
```

> 

Reviewer 3: @dockerymick add new step
Use Grafana on the new hub to verify that metrics are being collected from the managed clusters. Recent metrics should be visible for managed clusters within the last 5 minutes (default collection interval).

> . Shut down the Observability service on your primary hub cluster and clear in-memory metrics to S3 object after migrating all managed clusters. Run the following command:

Reviewer 3: After verifying that all ManagedClusters are connected to the new ACM hub and everything is functioning correctly, you can decommission the old ACM hub cluster by cleaning up its resources. To  shut down Observability on the primary hub, run the following command
```
"oc delete mco observability"
```

[#backup-restore-obs]
= Backing up and restoring Observability service

As you follow this procedure it is assumed that the same S3-compatible object store is used across both the primary and backup hub clusters to minimize disruption in metric data collection.

.Prerequsites

> - Some Observability resources must be manually backed up and restored to ensure continuity across hub clusters. Complete the following steps to manually backup required Observability resources:

Reviewer: this looks more like an intro line. (line 8)

> As you follow this procedure it is assumed that the same S3-compatible object store is used across both the primary and backup hub clusters to minimize disruption in metric data collection.
> 
> .Prerequsites
> 
> - Some Observability resources must be manually backed up and restored to ensure continuity across hub clusters. Complete the following steps to manually backup required Observability resources:
> 
> . To preserve the tenant ID of the `observatorium` resource during the restore, run the following command:
> 
> +
> [source,bash]
> ----
> oc get observatorium -n open-cluster-management-observability -o yaml > observatorium-backup.yaml
> ----

Author: ```suggestion
. Use the local-cluster renaming feature to assign unique names to each hub cluster. This helps determine metrics for each hub cluster in Grafana.

. To preserve the tenant ID of the `observatorium` resource during the restore, run the following command:

+
[source,bash]
----
oc get observatorium -n open-cluster-management-observability -o yaml > observatorium-backup.yaml
----
```

. To backup the `observability` deployment, run the following command:

+
[source,bash]
----
oc get mco observability -o yaml > mco-cr-backup.yaml
----

[#backup-restore-procedure]

> == Backup and restore for Observability

Reviewer: I think we can do without h2 here bc it is in the title.

Reviewer: @dockerymick and team, I recommend that we also add a link to these new topics from a general place in Obs. like an architecture topic or something like that bc an Obs user may not know where to find this.

[#backup-restore-obs]
= Backing up and restoring Observability service

Backup and restore the Observability service to maintain data continuity and integrity during hub cluster migration or backup. It is assumed that the same S3-compatible object store is used across both the primary and backup hub clusters to minimize disruption in metric data collection.

.Prerequsites

> - Enable self-management on your hub cluster to ensures its metrics are collected and visible in Grafana. The Observability service recognizes the hub cluster as a managed cluster.

Author: ```suggestion
- Any prereqs?
```

//need to look at the local cluster feature in adv config.^| B. Swope| during peer review
//are there prereqs? might also need help with the section title --- MJ

[#backup-restore-obs-procedure]
== Backing up and restoring Observability procedure

> . Manually back up and restore the `observatium` resource and `observability` deployment to ensure continuity across hub clusters.
> .. To preserve the tenant ID of the `observatorium` resource during the restore, run the following command:

Author: ```suggestion
. To ensure the Observability service recognizes the hub cluster as the `local-cluster`, change the `spec.disableHubSelfManagement` parameter in the `MultiClusterHub` custom resource to `false`. 
. Manually back up and restore the `observatium` resource and `observability` deployment to ensure continuity across hub clusters.
.. To preserve the tenant ID of the `observatorium` resource during the restore, run the following command:
```

Reviewer 3: Same logic goes for line 14. It is not a prerequisite.

[#backup-restore-obs]
= Backing up and restoring Observability service

Backup and restore the Observability service to keep data safe and to support continuity during the hub cluster migration or backup. To help with disruption in metric data collection, use the same S3-compatible object store for both the primary and backup hub clusters.

.Prerequsites

- Ensure that you can run a restore operation for backup types by completing the xref:../business_continuity/backup_restore/backup_restore.adoc#restoring-backup-restore-operation[Using the restore operation for backup types] process.

[#backup-restore-obs-procedure]

> == Backing up and restoring Observability procedure

Reviewer: We don't need this header. You can just use the procedure tag here because it is the same thing as the title.

Author: Ok thanks. I thought we were holding off on using tags, as in we would create a separate issue to handle the changes for the tool conversion. I am fine with using it now

+
[source,bash]
----
oc apply -f observatorium-backup.yaml
----

. Apply the backed up `MultiClusterObservability` custom resource to start the Observability service on the new restored hub cluster. Run the following command:

+
[source,bash]
----
oc apply -f mco-cr-backup.yaml
----
+
The operator starts the Observability service and detects the existing `observatorium` resource, reusing the preserved tenant ID instead of creating a new one.

. Verify that the Observability service runs on your new hub cluster. Run the following command:

+
[source,bash]
----
oc get pods -n open-cluster-management-observability
----

. Verify that the `observability-controller` `managedclusteraddon` does not have a status in the `DEGRADED` column, and that the `PROGRESSING` status is not set to `False`. Run the following command:

+
[source,bash]
----
oc get managedclusteraddons -A | awk 'NR==1 || /observability-controller/
----

. Verify metrics collection from your managed clusters by accesing Grafana.

. Verify that your managed clusters are connected to your new hub cluster by checking for the `Available` status for each managed cluster.

> . Shut down the Observability service on your old hub cluster by removing the resources. Run the following command:

Reviewer: use "previous" instead of "old"
